servicemanager/servicemanager_test.go

```diff
@@ -0,0 +1,545 @@
+	blobsDir := path.Join(imgFolder, "blobs")
+	if err := os.MkdirAll(blobsDir, 0o755); err != nil {
+		return digest, aoserrors.Wrap(err)
+	}
+
+	tarFile := path.Join(blobsDir, "_temp.tar.gz")
+
+	if output, err := exec.Command("tar", "-C", rootfs, "-czf", tarFile, "./").CombinedOutput(); err != nil {
+		return digest, aoserrors.New(fmt.Sprintf("error: %s, code: %s", string(output), err))
+	}
+	defer os.Remove(tarFile)
+
+	file, err := os.Open(tarFile)
+	if err != nil {
+		return digest, aoserrors.Wrap(err)
+	}
+	defer file.Close()
+
+	byteValue, err := ioutil.ReadAll(file)
+	if err != nil {
+		return digest, aoserrors.Wrap(err)
+	}
+
+	digest, err = generateAndSaveDigest(blobsDir, byteValue)
+	if err != nil {
+		return digest, aoserrors.Wrap(err)
+	}
+
+	os.RemoveAll(rootfs)
+
+	return digest, nil
+}
+
+func generateAndSaveDigest(folder string, data []byte) (retDigest digest.Digest, err error) {
+	fullPath := path.Join(folder, "sha256")
+	if err := os.MkdirAll(fullPath, 0o755); err != nil {
+		return retDigest, aoserrors.Wrap(err)
+	}
+
+	h := sha256.New()
+	h.Write(data)
+	retDigest = digest.NewDigest("sha256", h)
+
+	file, err := os.Create(path.Join(fullPath, retDigest.Hex()))
+	if err != nil {
+		return retDigest, aoserrors.Wrap(err)
+	}
+	defer file.Close()
+
+	_, err = file.Write(data)
+	if err != nil {
+		return retDigest, aoserrors.Wrap(err)
+	}
+
+	return retDigest, nil
+}
+
+func genarateImageManfest(folderPath string, imgConfig, aosSrvConfig, rootfsLayer *digest.Digest,
+	srvLayers []digest.Digest) (err error) {
+	type serviceManifest struct {
+		imagespec.Manifest
+		AosService *imagespec.Descriptor `json:"aosService,omitempty"`
+	}
+
+	var manifest serviceManifest
+	manifest.SchemaVersion = 2
+
+	manifest.Config = imagespec.Descriptor{
+		MediaType: "application/vnd.oci.image.config.v1+json",
+		Digest:    *imgConfig,
+	}
+
+	if aosSrvConfig != nil {
+		manifest.AosService = &imagespec.Descriptor{
+			MediaType: "application/vnd.aos.service.config.v1+json",
+			Digest:    *aosSrvConfig,
+		}
+	}
+
+	layerDescriptor := imagespec.Descriptor{
+		MediaType: "application/vnd.oci.image.layer.v1.tar+gzip",
+		Digest:    *rootfsLayer,
+	}
+
+	manifest.Layers = append(manifest.Layers, layerDescriptor)
+
+	for _, layerDigest := range srvLayers {
+		layerDescriptor := imagespec.Descriptor{
+			MediaType: "application/vnd.aos.image.layer.v1.tar",
+			Digest:    layerDigest,
+		}
+
+		manifest.Layers = append(manifest.Layers, layerDescriptor)
+	}
+
+	data, err := json.Marshal(manifest)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	jsonFile, err := os.Create(path.Join(folderPath, "manifest.json"))
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if _, err := jsonFile.Write(data); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func packImage(source, name string) (err error) {
+	log.WithFields(log.Fields{"source": source, "name": name}).Debug("Pack image")
```

Comment:
Why not use exec tar?

Reply:
was taken from the previous implementation. Should it be reworked?

Reply:
yes. If we use exec for untar what is the point to pack by tar, gzip writer?

Reply:
done